src/napari_builtins/_qt/features_table.py

```diff
@@ -0,0 +1,597 @@
+from __future__ import annotations
+
+from contextlib import contextmanager
+from pathlib import Path
+from typing import TYPE_CHECKING, Any
+
+import pandas as pd
+from qtpy.QtCore import (
+    QAbstractTableModel,
+    QItemSelection,
+    QItemSelectionModel,
+    QModelIndex,
+    QSortFilterProxyModel,
+    Qt,
+    QTimer,
+)
+from qtpy.QtGui import (
+    QGuiApplication,
+    QKeySequence,
+)
+from qtpy.QtWidgets import (
+    QAbstractItemView,
+    QComboBox,
+    QFileDialog,
+    QLabel,
+    QPushButton,
+    QStyledItemDelegate,
+    QTableView,
+    QVBoxLayout,
+    QWidget,
+)
+from superqt import QToggleSwitch
+
+from napari.utils.history import get_save_history
+from napari.utils.misc import in_ipython
+
+if TYPE_CHECKING:
+    import napari
+
+
+class PandasModel(QAbstractTableModel):
+    """Qt Model for a pandas DataFrame.
+
+    Follows the Qt Model/View protocol, implementing all the needed methods
+    for a QTableView to be able to read/write changes to a pandas dataframe.
+    Columns of specific dtypes get special treatment (bools become tickboxes,
+    categoricals are exposed through comboboxes, etc).
+    It's designed to be used in conjunction with the BoolFriendlyProxyModel
+    in order to properly sort, and with the DelegateCategorical in order to
+    provide comboboxes for categoricals.
+    """
+
+    def __init__(self, df: pd.DataFrame | None = None, parent=None):
+        super().__init__(parent)
+        self.df = df if df is not None else pd.DataFrame()
+        self.editable = False
+
+    # model methods necessary for qt
+    def rowCount(self, parent=None):
+        return self.df.shape[0]
+
+    def columnCount(self, parent=None):
+        return self.df.shape[1] + 1  # include index
+
+    def data(self, index: QModelIndex, role=Qt.ItemDataRole.DisplayRole):
+        if not index.isValid():
+            return None
+
+        row = index.row()
+        col = index.column()
+
+        if col == 0:  # index
+            if role in {Qt.ItemDataRole.DisplayRole, Qt.ItemDataRole.EditRole}:
+                return str(self.df.index[row])
+            return None
+
+        value = self.df.iat[row, col - 1]
+        dtype = self.df.dtypes.iat[col - 1]
+
+        # show booleans as respective checkboxes
+        if role == Qt.CheckStateRole and pd.api.types.is_bool_dtype(dtype):
+            return Qt.CheckState.Checked if value else Qt.CheckState.Unchecked
+
+        if role in {Qt.ItemDataRole.DisplayRole, Qt.ItemDataRole.EditRole}:
+            # format based on dtype
+            if pd.api.types.is_float_dtype(dtype):
+                return f'{value:.6g}'
+            if pd.api.types.is_integer_dtype(dtype):
+                return f'{value:d}'
+            if pd.api.types.is_datetime64_any_dtype(dtype):
+                return value.strftime('%Y-%m-%d')
```

Comment:
It will be nice to add tests that rendering and updating dates (from keyboard) works.